// browser-client/lib/browser-client.js
var _ = require('lodash');

module.exports = function() {
  return function(req, res, next) {
    var userAgent = req.headers['user-agent'];
    
    browserClient = {
      safari: false,
      firefox: false,
      chrome: false,
      ie: false,
      opera: false,
      operaMini: false,
      operaMobile: false,
      mobileSafari: false,
      
      browser: '',
      
      iphone: false,
      ipod: false,
      ipad: false,
      blackberry: false,
      
      device: '',
      
      mac: false,
      windows: false,
      linux: false,
      android: false,
      ios: false,
      
      os: ''
    };
    
    var version = /(?:Version|MSIE|Firefox|Chrome|QuickTime|BlackBerry[^\/]+|CoreMedia v)[\/ ]?([a-z0-9.]+)/i.exec(userAgent) || /Opera\/.*? Version\/([\d.]+)/.exec(userAgent);
    
    if(version) {
      browserClient.version = parseFloat(version[1]);
    }
    
    var testGroups = {
      browser: {
        chrome: function(ua) { return /Chrome/.test(ua); },
        firefox: function(ua) { return /Firefox/.test(ua); },
        ie: function(ua) { return /MSIE/.test(ua) && ! /Opera/.test(ua); },
        mobileSafari: function(ua) { return /Safari/.test(ua) && /(iPhone|iPad|iPod)/.test(ua); },
        safari: function(ua) { return /Safari/.test(ua); },
        operaMini: function(ua) { return /Opera Mini/.test(ua); },
        operaMobile: function(ua) { return /Opera Mobi/.test(ua); },
        opera: function(ua) { return /Opera/.test(ua); }
      },
      
      os: {
        mac: function(ua) { return /Mac OS X/.test(ua); },
        windows: function(ua) { return /Windows/.test(ua); },
        linux: function(ua) { return /Linux/.test(ua); },
        ios: function(ua) { return /(iPhone|iPod|iPad)/.test(ua) },
        android: function(ua) { return /Android/.test(ua); }
      },
      
      device: {
        iphone: function(ua) { return /iPhone/.test(ua); },
        ipad: function(ua) { return /iPad/.test(ua); },
        ipod: function(ua) { return /iPod/.test(ua); },
        blackberry: function(ua) { return /BlackBerry/.test(ua); }
      },
      
      platform: {
        mobile: function(ua, browserClient) {
          return (! (browserClient.mac || browserClient.windows || browserClient.linux) && ! browserClient.ipad && ! /Mobile/.test(ua));
        },
        
        tablet: function(ua, browserClient) {
          return (browserClient.ipad || browserClient.android);
        },
        
        desktop: function() {
          return true;
        }
      }
    };
    
    _.each(testGroups, function(tests, testKey){
      _.each(tests, function(fn, key){
        var result = fn(userAgent, browserClient);
        
        if(result) {
          browserClient[key] = true;
          browserClient[testKey] = key;
          return false;
        }
      });
    });
    
    req.browserClient = browserClient;
    
    next();
  };
};
